File: ch04/max_sub_arr.c
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>

#define CRASH { asm("ud2"); }

typedef struct SubArray {
    int loIdx;
    int hiIdx;
    int sum;
} SubArray;

SubArray *maxCrossingSubArray(int *arr, int loIdx, int midIdx, int hiIdx)
{
    int sum, leftSum, rightSum, maxLeft, maxRight;
    SubArray *subArr;

    subArr = malloc(sizeof(*subArr));

    leftSum = INT_MIN;
    sum = 0;
    for (int i = midIdx; i >= loIdx; --i) {
        sum += arr[i];
        if (sum > leftSum) {
            leftSum = sum;
            maxLeft = i;
        }
    }

    rightSum = INT_MIN;
    sum = 0;
    for (int j = midIdx + 1; j <= hiIdx; ++j) {
        sum += arr[j];
        if (sum > rightSum) {
            rightSum = sum;
            maxRight = j;
        }
    }

    subArr->loIdx = maxLeft;
    subArr->hiIdx = maxRight;
    subArr->sum = leftSum + rightSum;

    return subArr;
}

SubArray *maxSubArray(int *arr, int loIdx, int hiIdx)
{
    SubArray *subArr;

    // Edge case: single array element.
    if (hiIdx == loIdx) {
        subArr = malloc(sizeof(*subArr));
        subArr->loIdx = loIdx;
        subArr->hiIdx = hiIdx;
        subArr->sum = arr[loIdx];
        goto leave;
    }
    
    // Divide by 2 while keeping it an int.
    int midIdx = (loIdx + hiIdx) >> 1;

    SubArray *leftSubArr = maxSubArray(arr, loIdx, midIdx);
    SubArray *rightSubArr = maxSubArray(arr, midIdx + 1, hiIdx);
    SubArray *crossSubArr = maxCrossingSubArray(arr, loIdx, midIdx, hiIdx);

    if (leftSubArr->sum >= rightSubArr->sum
        && leftSubArr->sum >= crossSubArr->sum)
    {
        subArr = leftSubArr;
        free(rightSubArr);
        free(crossSubArr);
        goto leave;
    }
    else if (rightSubArr->sum >= leftSubArr->sum
        && rightSubArr->sum >= crossSubArr->sum)
    {
        subArr = rightSubArr;
        free(leftSubArr);
        free(crossSubArr);
        goto leave;
    }
    
    subArr = crossSubArr;
    free(leftSubArr);
    free(rightSubArr);

leave:
    return subArr;
}

int main(int argc, char **argv)
{
    int *arr, arrSz;
    SubArray *subArr;

    if (argc < 3) {
        printf("Must supply array.\n");
        exit(1);
    }

    arrSz = argc - 1;
    arr = malloc(arrSz*sizeof(*arr));
    for (int i = 0; i < arrSz; ++i) {
        arr[i] = atoi(argv[i + 1]);
    }

    subArr = maxSubArray(arr, 0, arrSz - 1);

    printf("[%d, %d, %d]\n", subArr->loIdx, subArr->hiIdx, subArr->sum);
    for (int i = subArr->loIdx; i <= subArr->hiIdx; ++i) {
        printf("%d\n", arr[i]);
    }

    free(subArr);
    free(arr);

    return 0;
}

File: ch04/max_sub_arr.py
import sys
import math

def main(argv):
    arr = [int(x) for x in argv[1:]]
    tmp = maxSubArray(arr, 0, len(arr) - 1)
    print(tmp)
    print(arr[tmp[0]:tmp[1] + 1])

def maxSubArray(arr, loIdx, hiIdx):
    if hiIdx == loIdx:
        return [loIdx, hiIdx, arr[loIdx]]
    else:
        midIdx = int(math.floor((loIdx + hiIdx) / 2))
        tmp = maxSubArray(arr, loIdx, midIdx)
        #print('left:\t' + str(tmp))
        leftLo = tmp[0]
        leftHi = tmp[1]
        leftSum = tmp[2]
        tmp = maxSubArray(arr, midIdx + 1, hiIdx)
        #print('right:\t' + str(tmp))
        rightLo = tmp[0]
        rightHi = tmp[1]
        rightSum = tmp[2]
        tmp = maxCrossingSubArray(arr, loIdx, midIdx, hiIdx)
        #print('cross:\t' + str(tmp))
        crossLo = tmp[0]
        crossHi = tmp[1]
        crossSum = tmp[2]
        if leftSum >= rightSum and leftSum >= crossSum:
            return [leftLo, leftHi, leftSum]
        elif rightSum >= leftSum and rightSum >= crossSum:
            return [rightLo, rightHi, rightSum]
        else:
            return [crossLo, crossHi, crossSum]

def maxCrossingSubArray(arr, loIdx, midIdx, hiIdx):
    leftSum = -sys.maxsize
    mySum = 0
    for i in range(midIdx, loIdx - 1, -1):
        mySum += arr[i]
        if mySum > leftSum:
            leftSum = mySum
            maxLeft = i
    rightSum = -sys.maxsize
    mySum = 0
    for j in range(midIdx + 1, hiIdx + 1):
        mySum += arr[j]
        if mySum > rightSum:
            rightSum = mySum
            maxRight = j
    return [maxLeft, maxRight, leftSum + rightSum]

if __name__ == '__main__':
    main(sys.argv)